File: tests/test_parse/data/sample.hs
module UserManager where

import Data.Maybe (fromMaybe)
import Control.Monad (when)
import qualified Data.Map as Map

-- Type definitions
data User = User
    { userName :: String
    , userAge :: Int
    } deriving (Show, Eq)

data Result a
    = Success a
    | Error String
    deriving (Show)

-- Type class
class Processable a where
    process :: a -> Result a

-- Type class instance
instance Processable User where
    process user@(User _ age)
        | age >= 18 = Success user
        | otherwise = Error "User is underage"

-- Basic function with type signature
createUser :: String -> Int -> User
createUser name age = User name age

-- Pattern matching function
getUserName :: User -> String
getUserName (User name _) = name

-- Guards in function
checkAge :: Int -> String
checkAge age
    | age < 0 = "Invalid age"
    | age < 18 = "Minor"
    | age < 65 = "Adult"
    | otherwise = "Senior"

-- Function with where clause
calculateStats :: User -> (Int, String)
calculateStats user = (ageCategory, status)
    where
        age = userAge user
        ageCategory = age `div` 10 * 10
        status = checkAge age

-- Function using let expression
formatUser :: User -> String
formatUser user =
    let name = userName user
        age = userAge user
    in name ++ " (" ++ show age ++ " years old)"

-- Higher order function
processUsers :: (User -> Result User) -> [User] -> [Result User]
processUsers f = map f

-- Function composition
getProcessedName :: User -> String
getProcessedName = formatName . userName
    where formatName = reverse . map toUpper

-- Curried function
addAge :: Int -> User -> User
addAge years user = user { userAge = userAge user + years }

-- Maybe returning function
findUser :: String -> [User] -> Maybe User
findUser name = foldr check Nothing
    where check user acc
            | userName user == name = Just user
            | otherwise = acc

-- List comprehension
getAdultUsers :: [User] -> [User]
getAdultUsers users = [user | user <- users, userAge user >= 18]

-- Function using case expression
processUserCase :: User -> Result User
processUserCase user = case userAge user of
    age | age < 0 -> Error "Invalid age"
        | age < 18 -> Error "Underage"
        | otherwise -> Success user

-- Partial application example
incrementAge :: User -> User
incrementAge = addAge 1

-- Lambda function example
processWithLog :: User -> IO ()
processWithLog = \user -> do
    putStrLn $ "Processing user: " ++ userName user
    print $ process user

-- Main function to demonstrate usage
main :: IO ()
main = do
    let user1 = createUser "John" 25
        user2 = createUser "Alice" 17
        users = [user1, user2]
    
    putStrLn "Created users:"
    mapM_ (putStrLn . formatUser) users
    
    putStrLn "\nProcessing users:"
    mapM_ (print . process) users
    
    putStrLn "\nAdult users:"
    mapM_ (putStrLn . formatUser) (getAdultUsers users)
    
    putStrLn "\nProcessing with logs:"
    mapM_ processWithLog users
    
    let olderUser1 = incrementAge user1
    putStrLn $ "\nIncremented age for " ++ formatUser olderUser1 

File: tests/test_parse/data/sample.ml
(* Module signature *)
module type UserManager = sig
  type user
  type 'a result = Ok of 'a | Error of string
  
  val create_user : string -> int -> user
  val process_user : user -> user result
  val get_name : user -> string
  val get_age : user -> int
  val format_user : user -> string
end

(* Module implementation *)
module User : UserManager = struct
  type user = {
    name: string;
    age: int;
  }

  type 'a result = Ok of 'a | Error of string

  (* Basic function *)
  let create_user name age =
    { name = name; age = age }

  (* Pattern matching function *)
  let process_user user =
    match user.age with
    | age when age < 0 -> Error "Invalid age"
    | age when age < 18 -> Error "Underage"
    | _ -> Ok user

  (* Record field accessor functions *)
  let get_name user = user.name
  let get_age user = user.age

  (* String formatting function *)
  let format_user user =
    Printf.sprintf "%s (%d years old)" user.name user.age
end

(* Function with labeled arguments *)
let update_user ~name ~age user =
  { user with User.name = name; age = age }

(* Higher-order function *)
let process_users f users =
  List.map f users

(* Function with optional argument *)
let create_greeting ?(prefix="Hello") name =
  prefix ^ ", " ^ name ^ "!"

(* Recursive function *)
let rec find_user name = function
  | [] -> None
  | user :: rest ->
      if User.get_name user = name then Some user
      else find_user name rest

(* Function using List module *)
let get_adult_users users =
  List.filter (fun user -> User.get_age user >= 18) users

(* Partial application *)
let is_adult = (>=) 18

(* Function composition *)
let get_uppercase_name user =
  user |> User.get_name |> String.uppercase_ascii

(* Function with local bindings *)
let process_and_format user =
  let open User in
  let result = process_user user in
  match result with
  | Ok processed -> "Success: " ^ format_user processed
  | Error msg -> "Error: " ^ msg

(* Curried function *)
let add_ages user1 user2 =
  User.get_age user1 + User.get_age user2

(* Main execution *)
let () =
  let user = User.create_user "John" 25 in
  Printf.printf "%s\n" (User.format_user user)

(* Example OCaml module *)
module User = struct
  type t = {
    name: string;
    age: int;
  }

  let create name age = { name; age }
  
  let is_adult user = user.age >= 18
  
  let to_string user =
    Printf.sprintf "%s (%d years old)" user.name user.age
end

let () =
  let user = User.create "John" 25 in
  Printf.printf "%s\n" (User.to_string user) 